torch/csrc/jit/passes/subgraph_matcher.cpp

```diff
@@ -0,0 +1,269 @@
+#include <torch/csrc/jit/passes/subgraph_matcher.h>
+
+namespace torch {
+namespace jit {
+
+MatchIterator& MatchIterator::begin() {
+  tried_nodes_.clear();
+  updateMatch();
+  return *this;
+}
+
+MatchIterator& MatchIterator::end() {
+  return *this;
+}
+
+MatchIterator& MatchIterator::operator++() {
+  updateMatch();
+  return *this;
+}
+
+MatchIterator& MatchIterator::operator++(int) {
+  updateMatch();
+  return *this;
+};
+
+MatchIterator::Match MatchIterator::operator*() {
+  return match_;
+}
+
+bool MatchIterator::operator!=(MatchIterator& rhs) {
+  if (finished_)
+    return false;
+  return true;
+}
+
+const Node* MatchIterator::findNewNode() {
+  for (const auto node : graph_.nodes()) {
+    if (tried_nodes_.count(node) || !matchNodes(anchor_, node)) {
+      continue;
+    }
+    tried_nodes_.insert(node);
+    current_block_ = node->owningBlock();
+    return node;
+  }
+  return nullptr;
+}
+
+// Macro to help with match requirements
+#define REQ(cond)                \
+  if (!(cond)) {                 \
+    log("[mismatch] cond failed " #cond); \
+    return false;                \
+  } else { \
+    log("[match] cond " #cond); \
+  }
+
+bool MatchIterator::matchNodeFrontier(
```

Comment:
Could you please add some comments about how the algorithm works and what the frontiers are?